### ISSUE TITLE
tsc --watch regression in Typescript 1.8 Beta

### ISSUE THREAD

**Issue by Jason-Rev** [Microsoft/TypeScript#6941]
With the beta version of 1.8 installed using:

```
npm install typescript@1.8.0 -g
```
tsc --watch from project root does not appear to be working.  It compiles once but any subsequent changes to the files has no effect.

going back to 1.7.5, everything works fine.

This seems related to [#6511](https://github.com/Microsoft/TypeScript/issues/6511).

npm install typescript@next which gets version 1.9.0 also does not seem to be working.

**Comment by mhegazy** [comment]
What version of node are you running? What os? And are any files that are watched correctelly?

**Comment by Jason-Rev** [comment]
Good questions:
Node: v5.1.1
OS X 10.10.5
Bash terminal
It seems to be an issue with both .tsx and .ts files.
I wasn't able to get it to detect any changes.

**Comment by Jason-Rev** [comment]
Upgrading to Node: v5.5.0 didn't seem to make a difference.

**Comment by mhegazy** [comment]
@billti do you see a similar behavior? i do not see issues on v.5.5.0 on windows.

**Comment by billti** [comment]
I just installed Node v5.5.0 and created a simple project with no issues. I added a .tsx file, tried with a couple of different "jsx" settings, and `tsc -w` picked up changes and compiled to my `outDir` fine.

I'm using the latest OS X (10.11.3) but it seems unlikely this is the cause. Anything beyond a very typical setup that might be causing this? If you create a new super-simple project with only a couple of files do you see the same problem?

**Comment by Jason-Rev** [comment]
I don't have an outDir specified:

```javascript
{
  "compilerOptions": {
    "target": "ES5",
    "module": "commonjs",
    "moduleResolution": "node",
    "sourceMap": true,
    "jsx": "react",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": false,
    "noImplicitAny": false,
    "preserveConstEnums": true
  },
  "exclude": [
    "node_modules",
    "bower_components"
  ]
}
```

I'll see if I can create a small sample project.

**Comment by Jason-Rev** [comment]
I have made a project that seems to break: [tsc-watch](https://github.com/Jason-Rev/tsc-watch)
Note: It is not a working project, is just demonstrates the --watch issue.

Install:
```
git clone git@github.com:Jason-Rev/tsc-watch.git
cd tsc-watch
npm install
tsd install
tsc -w
```
Everything should build without warnings or errors.
.js and .map files will be created next to the .ts and .tsx files.

Choose test/test.ts file to edit.
Add a few lines and save.
For me, the changes to test.ts are not picked up.  Changes to other files are also not picked up.

Cheers,
Jason

**Comment by mhegazy** [comment]
This project is working fine for me on my Windows box. @vladima can you give tsc--w on Ubuntu a try as well.

**Comment by Jason-Rev** [comment]
To be clear, tsc -w picks up new files.  But it does not pick up changes to existing files.

**Comment by vladima** [comment]
Ubuntu 15.10
- cloned project, `npm install & ./node_modules/.bin/tsd install`
- `./node_modules/.bin/tsc -w -p .` - tsc is started in watch mode
- edited existing files `test/test.ts` and `test/actions/actions.ts` - all changes were picked up after file is saved
- added new file `newFile.ts` at the project root, edited it - worked fine, project is recompiled after saving file

Please let me know if anything is missed or you've done it differently.

**Comment by Jason-Rev** [comment]
I wonder if it is a case-sensitive issue.  OSX is case-insensitive.  I believe Ubuntu is case-sensitive.

I had a brief look at the code that changed around the watcher and it looked like there were different code paths for case-sensitive vs case-insensitive.

**Comment by mhegazy** [comment]
#6511 was a change related to case sensitivity. but that was fixed a while ago. we have not touched this code since. do you see this on other machines? is it all projects? does it happen with no tsconfig.jon (e.g. tsc --w file.ts)?

**Comment by geowarin** [comment]
Same problem here on OSX 10.11.3.
Watch does not pick up any change neither with a tsconfig file nor listing the files in the commandline.
I reproduced the issue with `typescript@1.8.0-dev.20160125` and `typescript@1.9.0-dev.20160209`.

**Comment by Jason-Rev** [comment]
I found the issue.
I think it broke when the watch logic between directories and files was separated.

The issue is with how changes are tracked in:  sys.ts:fileEventHandler

Here is why I see it and other people are not.
I am using WebStorm as an editor.
When it saves changes to a file, it does not write directly to the file.  Instead it:

1. writes to ```test.ts___jb_bak___``` 
2. it then renames ```test.ts``` to ```test.ts___jb_old___```.  
3. Once than is done, it renames ```test.ts___jb_bak___``` to ```test.ts```
4. Because ```fileEventHandler``` is only looking at "change" events, it misses out on the fact the file it was watching got renamed and replaced by a new file.

sys.ts
```javascript
function fileEventHandler(eventName: string, relativeFileName: string, baseDirPath: Path) {
    // When files are deleted from disk, the triggered "rename" event would have a relativefileName of "undefined"
    const filePath = typeof relativeFileName !== "string"
        ? undefined
        : toPath(relativeFileName, baseDirPath, createGetCanonicalFileName(sys.useCaseSensitiveFileNames));
    if (eventName === "change" && fileWatcherCallbacks.contains(filePath)) {
        for (const fileCallback of fileWatcherCallbacks.get(filePath)) {
            fileCallback(filePath);
        }
    }
}
```

the fix is to correctly handle "rename" events.

Simple Repro steps:
1. Use the given project
2. ```tsc -w``` in the project root
3. In a new shell ```cd test```
4. ```cat test.ts > test.ts_new```
5. ```echo "\n// changes" >> test.ts_new```
6. ```mv test.ts test.ts_old```
7. ```mv test.ts_new test.ts```

Expected that the new ```test.ts``` file get compiled.
Actual: nothing happens.

**Comment by geowarin** [comment]
@Jason-Rev Good catch! I'm using IntelliJ.
Changes are indeed correctly picked up when I save the file in a text editor.

**Comment by geowarin** [comment]
Trivially changing the test to

```javascript
(eventName === "change" || eventName == "rename")
```

seems to work but also produces a scary error:

```
2016-02-09 16:03 node[79800] (FSEvents.framework) FSEventStreamFlushSync(): failed assertion '(SInt64)last_id > 0LL'
4:03:40 PM - File change detected. Starting incremental compilation...
```

**Comment by billti** [comment]
Looks like a very lengthy write-up of the cause of that exception at https://github.com/bdkjones/fseventsbug/wiki/realpath()-And-FSEvents . (Found by a reference to it on a related Node issue - https://github.com/nodejs/node/issues/854 ).

Will did into this a little more and see what the best solution might be...

**Comment by billti** [comment]
I've spent a couple hours on this, and I've been unable to repro the *scary* `FSEvent` error when handling the `rename` operations. The article above does state this is fixed in the latest OS X, which I'm on, so perhaps that's why I can't reproduce it. (Though @geowarin , you state you are on OS X 10.11.3, which is the same as me. Can you confirm you are seeing the FSEvent error on this OS version)?

There was also a bug in how `--watch` was detecting the compiler options if using a config file, which caused me a bunch of time, but I fixed that also. Pushing the pull request as soon as the tests pass...

**Comment by geowarin** [comment]
@billti Yeah, the `FSEventStreamFlushSync` is probably unrelated. I get it sometimes with typescript in watch mode. I do have the latest OSX version so there is still a bug out there :smile: 

Reading the article you linked to, it seems to be a random problem with the filesystem so typescript should probably not worry about it. Moreover, the compiler seem to work fine even with the error message showing. Sorry if it has caused confusion.

**Comment by Jason-Rev** [comment]
Thank you @billti.  I'll verify that your fix worked and close later today.

**Comment by billti** [comment]
Thanks. This is merged into release-1.8 and master now. Add a comment if you still see any issues (other than the benign OS X message) and I'll reopen.
